community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/staging/ForkedProcessorStep.java

```diff
@@ -222,7 +237,6 @@ public void run()
                     {
                         sendDownstream( candidate );
                     }
-                    current.next = null;
```

Comment:
The reason that this was here is that the 'noop' list entry was the very head of the list was kept as a field, keeping the entire list in memory. We fixed that and this call was no longer needed.

The reason that we wanted to remove this line, however, is that it could race with threads that had fallen behind, rendering them unable to catch up & causing a deadlock.

Reply:
You are saying it's possible for `current.isCompleted() == false && current.next.isCompleted() == true`? That means ForkedProcessors can complete their assigned work out-of-order, which is a little surprising.

Reply:
The reason we don't null this anymore is that not all units will be processed by all processors and so some units will be completed before all processors have seen them. And to allow those processors to still follow the chain it'll have to be remain.

Reply:
Or maybe this can be changed since changing number of processors awaits and depletes the current queue of units... hmm

Reply:
The main takeaway is, probably, since we now can have idle threads many batches can be finished while the excess threads are resting. If they are needed later and need to catch up they need this path in order to find the active part of the list.

Reply:
Alright, let's keep it as is for now. Especially since new processors now catch up from the tail.